add workflow activation and scheduling lock TTL

garden-of-delete opened the pull request.

garden-of-delete commented on service/scheduler.go:
```diff
@@ -59,18 +59,38 @@ type Scheduler struct {
-		fmt.Println("tick")
-		s.scheduleWorkflows(database.GetInstance())
-		s.activateWorkflows(database.GetInstance())
+	lockTtlTick := time.Tick(time.Minute * 5) // TODO: ok to hardcode to 5 mins?
```
Hardcoded to check for and delete locks older than LockTTL every 5 minutes. Could also do it every second, or parameterize the interval in the config.

garden-of-delete replied:
Moved the default lockTTL up to 15 mins per our convo

garden-of-delete commented on cmd/scheduler.go:
```diff
@@ -82,4 +85,11 @@ func init() {
+
+	schedulerCmd.Flags().Duration(
+		"lockTTL",
+		time.Hour,
```
is one hour a good default duration for the LockTTL?

schowsf replied:
I think more flexible to have in units of time.Minute

garden-of-delete replied:
Agreed, done.

garden-of-delete commented on service/scheduler.go:
```diff
@@ -59,18 +59,38 @@ type Scheduler struct {
 	}
 }
 
+func (s *Scheduler) deleteOldLocks(db *gorm.DB) {
```
handles the case where the pod crashes before the lock can be released in the lockAndCreate / lockAndActivate methods

garden-of-delete commented on service/scheduler.go:
```diff
@@ -59,18 +59,38 @@ type Scheduler struct {
-		fmt.Println("tick")
-		s.scheduleWorkflows(database.GetInstance())
-		s.activateWorkflows(database.GetInstance())
+	lockTtlTick := time.Tick(time.Minute)
```
In this implementation, `lockTtlTick` is the tick that defines the frequency that the scheduler checks for expired locks. LockTTL is a separate value that defines when a lock is considered expires.

garden-of-delete commented on service/scheduler.go:
```diff
@@ -59,18 +59,38 @@ type Scheduler struct {
+			fmt.Println("tick")
+			s.scheduleWorkflows(database.GetInstance())
+			s.activateWorkflows(database.GetInstance())
+		case <-lockTtlTick:
```
I considered that option. Just wasn't sure if we wanted to be checking for and removing expired locks every time the scheduler `lockAndCreate`s. I can move it and remove the separate ticker for that check.

garden-of-delete commented:
Moved the "check for and delete expired locks" logic into the `lockAndCreate` method. Also refactored `lockTTL` to `lockTimeout` and `deleteOldLocks` to `deleteExpiredLocks`.
